lib/ts/IntrusiveDList.h

```diff
@@ -43,339 +43,553 @@
 
 /// FreeBSD doesn't like just declaring the tag struct we need so we have to include the file.
 #include <iterator>
+#include <type_traits>
 
 /** Intrusive doubly linked list container.
 
-    This holds items in a doubly linked list using members of the
-    items.  Elements are copied in to the list. No memory management
-    is done by the list implementation.
+    This holds items in a doubly linked list using links in the items. Items are placed in the list by changing the
+    pointers. An item can be in only one list for a set of links, but an item can contain multiple sets of links. This
+    requires different specializations of this template because link access is part of the type specification. Memory
+    for items is not managed by this class - instances must be allocated and released elsewhere. In particular removing
+    an item from the list does not destruct or free the item.
 
-    To use this class a client should create the structure for
-    elements of the list and ensure that it has two self pointers to
-    be used by the list. For example,
+    Access to the links is described by a linkage class which is required to contain the following members:
+
+    - The static method @c next_ptr which returns a reference to the pointer to the next item.
+
+    - The static method @c prev_ptr which returns a reference to the pointer to the previous item.
+
+    The pointer methods take a single argument of @c Item* and must return a reference to a pointer instance. This
+    type is deduced from the methods and is not explicitly specified. It must be cheaply copyable and stateless.
+
+    An example declaration woudl be
 
     @code
-      struct Elt {
-        int _payload;
-        Elt* _next;
-        Elt* _prev;
+      // Item in the list.
+      struct Thing {
+        Thing* _next;
```

Comment:
Do these members have to be publicly accessible? It doesn't look like a beautiful C++ code to me.

Reply:
No, they don't need to be public, if the `Linkage` structure is marked as as `friend`. I do it that way because that is how the old "DLL" classes and links work. This version is more flexible and doesn't require it.

Another point of using the `Linkage` structure is that links declared in a super class can be used and the `Linkage` can do the internal `static_cast` to make it work (e.g. the actual link type would be `super_type*` but the accessor methods can cast that down to `self_type*`).

Reply:
Since I had to fix the clang formatting, I added tests to demonstrate that works.

Reply:
Since `Linkage` is an inner class, doesn't it work without a `friend` declaration?

Reply:
Possibly, but that's not been my experience. I don't think inner classes get privileged access.

Reply:
Apparently the `friend` isn't required. That must have changed at some point. I'll have to make note of it.

Reply:
Since C++11.
https://en.cppreference.com/w/cpp/language/nested_types